rs/cli/src/ic_admin.rs

```diff
@@ -679,112 +785,116 @@ must be identical, and must match the SHA256 from the payload of the NNS proposa
         let removed_entries: BTreeMap<usize, &FirewallRule> =
             rules.into_iter().filter(|(key, _)| !edited.contains_key(key)).collect();
 
-        fn pretty_print_diff(change_type: &str, entries: &BTreeMap<usize, &FirewallRule>) {
-            if entries.is_empty() {
-                return;
-            }
-            let diff = serde_json::to_string_pretty(entries).unwrap();
-            info!("Pretty printing diff ----- {}:\n{}", change_type, diff);
+        let mut mods = FirewallRuleModifications::new();
+        for (pos, rule) in added_entries.into_iter() {
+            mods.addition(pos, rule.clone());
+        }
+        for (pos, rule) in updated_entries.into_iter() {
+            mods.update(pos, rule.clone());
+        }
+        for (pos, rule) in removed_entries.into_iter() {
+            mods.removal(pos, rule.clone());
         }
 
-        pretty_print_diff("added", &added_entries);
-        pretty_print_diff("updated", &updated_entries);
-        pretty_print_diff("removed", &removed_entries);
-
-        if added_entries.is_empty() && updated_entries.is_empty() && removed_entries.is_empty() {
+        let reverse_sorted = mods.reverse_sorted_and_batched();
+        if reverse_sorted.is_empty() {
             info!("No modifications should be made");
             return Ok(());
         }
-
-        if added_entries.len() + updated_entries.len() + removed_entries.len() > 1 {
+        let diff = serde_json::to_string_pretty(&reverse_sorted).unwrap();
+        info!("Pretty printing diff:\n{}", diff);
+        /*if reverse_sorted.len() > 1 {
             return Err(anyhow::anyhow!(
-                "Cannot apply more than 1 change at a time due to hash changes"
+                "Cannot currently apply more than 1 change at a time due to hash changes"
             ));
-        }
+        }*/
```

Comment:
remove?

Reply:
I'll add that this change does not impact any critical piece of configuration.  This change merely modifies how we call `ic-admin` and what data we pass onto it.  I want to be very clear about the scope of this change:

There are two ways in which someone could maliciously or accidentally operate this tool in which it could hypothetically break the IC.  The operator -- or the tool -- could enter an invalid payload (junk), or he could enter a vexatious yet valid payload (e.g. "block all connections from all IPs")

W.r.t invalid payloads:

* The correct place to establish safeguards as to what critical data and code is valid is on the server side (the replica), and the safeguards are already established there plus they are well-tested.
* The second place where safeguards should be in place is `ic-admin` itself (since `dre` merely invokes it here).  I am happy to report that I have personally tested `ic-admin` refuses to load invalid payloads, so the safeguards are already in place.

W.r.t vexatious payloads:

Nothing we can do from the perspective of maintainers of `dre`.  If someone sends a proposal that wrecks the IC, and the proposal gets voted in, then that's not a responsibility of the authors of the `dre` tool.

We could hypothetically add some heuristic that prevents extreme breakage (detect overly zealous block rules, like `rm -rf /` does today).  That, however, would not make this specific change any safer -- it is an engineering decision entirely orthogonal to this change.